x-pack/plugins/apm/server/lib/services/annotations/get_derived_service_annotations.ts

```diff
@@ -73,52 +75,140 @@ export async function getDerivedServiceAnnotations({
     return [];
   }
   const annotations = await Promise.all(
-    versions.map(async (version) => {
-      const response = await apmEventClient.search(
-        'get_first_seen_of_version',
-        {
-          apm: {
-            events: [
-              getProcessorEventForAggregatedTransactions(
-                searchAggregatedTransactions
-              ),
-            ],
+    versions.map((version) =>
+      serviceVersionPerServiceNode
+        ? getFirstSeenVersionByServiceNode(
+            version,
+            apmEventClient,
+            start,
+            end,
+            filter,
+            searchAggregatedTransactions
+          )
+        : getFirstSeenVersion(
+            version,
+            apmEventClient,
+            start,
+            end,
+            filter,
+            searchAggregatedTransactions
+          )
+    )
+  );
+  return annotations.flat().filter(Boolean) as Annotation[];
+}
+
+async function getFirstSeenVersion(
+  version: string,
+  apmEventClient,
+  start: number,
+  end: number,
+  filter: ESFilter[],
+  searchAggregatedTransactions: boolean
+) {
+  const response = await apmEventClient.search('get_first_seen_of_version', {
+    apm: {
+      events: [
+        getProcessorEventForAggregatedTransactions(
+          searchAggregatedTransactions
+        ),
+      ],
+    },
+    body: {
+      size: 1,
+      query: {
+        bool: {
+          filter: [...filter, { term: { [SERVICE_VERSION]: version } }],
+        },
+      },
+      sort: {
+        '@timestamp': 'asc',
+      },
+    },
+  });
+
+  const firstSeen = new Date(
+    response.hits.hits[0]._source['@timestamp']
+  ).getTime();
+
+  if (!isFiniteNumber(firstSeen)) {
+    throw new Error(
+      'First seen for version was unexpectedly undefined or null.'
+    );
+  }
+
+  if (firstSeen < start || firstSeen > end) {
+    return [];
+  }
+
+  return [
+    {
+      type: AnnotationType.VERSION,
+      id: version,
+      '@timestamp': firstSeen,
+      text: version,
+    },
+  ];
+}
+
+async function getFirstSeenVersionByServiceNode(
+  version: string,
+  apmEventClient,
+  start: number,
+  end: number,
+  filter: ESFilter[],
+  searchAggregatedTransactions: boolean
+) {
+  const response = await apmEventClient.search('get_first_seen_of_version', {
```

Comment:
the id for this query should be different. Like: `get_first_seen_of_version_by_service_node`

Reply:
Thank You @ogupte for your review. I changed the the id of the new query.